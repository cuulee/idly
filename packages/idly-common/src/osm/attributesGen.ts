import { Attributes } from '../osm/structures';
const attr = {};

export function attributesGen(
  {
    visible,
    version,
    timestamp,
    changeset,
    uid,
    user
  }: {
    visible?: boolean;
    version?: string;
    timestamp?: string;
    changeset?: string;
    uid?: string;
    user?: string;
  } = {}
): Attributes {
  return {
    changeset,
    timestamp,
    uid,
    user,
    version,
    visible
  };
}
