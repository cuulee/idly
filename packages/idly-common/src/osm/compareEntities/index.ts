export { compareNode } from './compareNode';
export { compareRelation } from './compareRelation';
export { compareTags } from './compareTags';
export { compareWay } from './compareWay';
