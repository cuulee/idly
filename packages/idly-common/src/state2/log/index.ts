import { weakCache } from '../../misc/weakCache';
import { setCreate, setFind, setIntersect, setSome } from '../helper';
/**
 * The log is used as [...modifiedEntries, virginOsmEntries]
 * the pointer to virginOsmEntries resolves to EntityTable
 * we carefully mutate virginOsmEntries keeping the pointer same.
 *
 * modified entities are id#version where version [0, infinity) (0 is inclusive)
 */

export type Index = string;

export type Log = ReadonlyArray<Entry>;

export type Entry = ReadonlySet<string>;

export const validId = (str: string) => str.indexOf('#') === 0;

export const validIndex = (str: string) => !validId(str);

export const logGetModifiedIds = weakCache((log: Log) =>
  log.reduce((prev: Set<string>, cur) => {
    cur.forEach(index => prev.add(getId(index)));
    return prev;
  }, setCreate())
);

export const logGetLatestIndexes = weakCache(
  (log: Log) =>
    new Set(
      log
        .reduce((prev, cur) => {
          cur.forEach(
            index => !prev.has(getId(index)) && prev.set(getId(index), index)
          );
          return prev;
        }, new Map<string, string>())
        .values()
    )
);

export const logRecreate = (...entries: Entry[]) =>
  entries
    .slice(0)
    .reverse()
    .reduce((l, e) => {
      const newLog = addEntryToLog(e)(l);
      if (l === newLog) {
        throw new Error('malformed entries');
      }
      return newLog;
    }, logCreate());

export const logCreate = (): Log => Object.freeze([]);

export const doesLogContain = (entry: Entry) => (log: Log) =>
  log.some(e => setIntersect(entry, e).size > 0);

export const addEntryToLog = (entry: Entry) => (log: Log) =>
  setSome(index => doesLogHaveNewerOrCurrentVersion(index)(log), entry)
    ? log
    : Object.freeze([entry, ...log]);

export const doesLogHaveNewerOrCurrentVersion = (index: string) => (log: Log) =>
  getLatestVersion(index)(log) >= getVersion(index);

export const getVersion = (index: Index) => parseInt(parseIndex(index)[1], 10);

export const getLatestVersion = (id: string) => (log: Log) => {
  const finder = findInEntry(getId(id));
  let index;
  for (const entry of log) {
    index = finder(entry);
    if (index) {
      return getVersion(index);
    }
  }
  return -1;
};

export const genIndex = (id: string) => (log: Log) =>
  `${getId(id)}#${getLatestVersion(id)(log) + 1}`;

export const findInEntry = (indexOrId: string) => (entry: Entry) =>
  setFind(i => getId(i) === indexOrId, entry);

export const getId = (index: Index) => parseIndex(index)[0];

const parseIndex = (index: Index) => index.split('#');

export const incrVersion = (index: Index) =>
  `${getId(index)}#${getVersion(index) + 1}`;
