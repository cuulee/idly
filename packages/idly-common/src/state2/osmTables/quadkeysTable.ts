import { iterableFlattenToSet, mapFilter, setCreate } from '../helper';
import {
  OneToManyTable,
  oneToManyTableCreate,
  oneToManyTableFilter,
  oneToManyTableUpdateIndex,
  tableFindKey,
  tableForEach,
  tableRemove,
} from '../table';
import { tableAdd, tableValues } from '../table/regular';

export type QuadkeysTable = OneToManyTable<string>;

/**
 * All the modifiedIds are stored in the
 * global root '' quadkey, the virgin Ids are always
 * stored in whatever quadkey they came from. The
 * advantage being modifiedIds being available for render
 * at all times.
 */

export function quadkeysTableAdd(
  t: QuadkeysTable,
  entityIds: string[],
  quadkey: string
) {
  if (ancestorProperFind(t, quadkey)) {
    console.log(
      'warning, ' +
        quadkey +
        'already has an ancestor =' +
        ancestorProperFind(t, quadkey)
    );
    return;
  }

  // since root quadkey is root of all, we just
  //  to append modifiedIds to the root quadkey
  if (quadkey === '') {
    oneToManyTableUpdateIndex(t, quadkey, entityIds);
    return;
  }

  // removes any descendant in favour of the parent quadkey

  removeAllDescendants(t, quadkey);

  tableAdd(t, quadkey, setCreate(entityIds));
}

export function removeAllDescendants(t: QuadkeysTable, quadkey: string) {
  tableForEach((_, id) => {
    if (isQuadkeyDescendant(quadkey, id)) {
      tableRemove(t, id);
    }
  }, t);
}

export const quadkeysTableCreate = (): QuadkeysTable => oneToManyTableCreate();

// @TOFIX with our removeAllDescendants here I think we will always have
// a flat system of quadkeys so this function can simplified to just returning
// a Map with a single quakey & the '' key. But we need to make sure we dont
// add a descendant, because removeAllDescendants will not remove any parent.
export const quadkeysTableCreateFrom = (t: QuadkeysTable, quadkey: string) =>
  mapFilter((_, k) => isQuadkeyRelated(quadkey, k), t);

export const quadkeysTableFlatten = (t: QuadkeysTable) =>
  iterableFlattenToSet(tableValues(t));

// Finds the virginIds assuming all modified Ids stay in '' quadkey
export const quadkeysTableFindVirginIds = (
  t: QuadkeysTable,
  quadkeys: string[]
): Set<string> =>
  quadkeys.some(q => q === '')
    ? setCreate()
    : iterableFlattenToSet(
        oneToManyTableFilter(
          (_, k) => k !== '' && quadkeys.some(q => isQuadkeyRelated(q, k)),
          t
        )
      );

export const isQuadkeyRelated = (q: string, maybeRelated: string) =>
  q === maybeRelated ||
  isQuadkeyDescendant(q, maybeRelated) ||
  isQuadkeyAncestor(q, maybeRelated);

export const isQuadkeyDescendant = (q: string, maybeDescendant: string) =>
  maybeDescendant.startsWith(q) && q !== maybeDescendant;

export const isQuadkeyAncestor = (q: string, maybeAncestor: string) =>
  q.startsWith(maybeAncestor) && q !== maybeAncestor;

export const isQuadkeyParent = (q: string, maybeParent: string) =>
  maybeParent === q.slice(0, -1);

export const isQuadkeyChild = (q: string, maybeChild: string) =>
  q === maybeChild.slice(0, -1);

// excludes '' as an ancestor
export const ancestorProperFind = (t: QuadkeysTable, quadkey: string) =>
  tableFindKey((_, id) => isQuadkeyAncestor(quadkey, id) && id !== '', t);
