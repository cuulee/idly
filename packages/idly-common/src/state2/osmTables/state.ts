import { Entity } from '../../osm/structures';
import {
  getAllLatestIndexes,
  Log,
  logCreate,
  logVirginIdsCurrentlyModified,
} from '../log';
import { Table } from '../table';
import { tableFilter } from '../table/regular';
import {
  Element,
  ElementTable,
  elementTableBulkAdd,
  elementTableCreate,
} from './elementTable';
import {
  QuadkeysTable,
  quadkeysTableAdd,
  quadkeysTableCreate,
  quadkeysTableCreateFrom,
  quadkeysTableFindVirginIds,
  quadkeysTableFlatten,
} from './quadkeysTable';

// TOFIX _elementTable & _quadkeysTable
// ref will change when going to a new state
// we need to put deep inside an object
// also write a test for this
export class State {
  public static create(
    { elementTable, quadkeysTable, log } = {
      elementTable: elementTableCreate(),
      log: logCreate(),
      quadkeysTable: quadkeysTableCreate(),
    }
  ) {
    return new State(elementTable, quadkeysTable, log);
  }

  private readonly log: Log;

  // tslint:disable-next-line:variable-name
  private _elementTable: Table<Element>;
  // tslint:disable-next-line:variable-name
  private _quadkeysTable: QuadkeysTable;

  private constructor(
    elementTable: ElementTable,
    quadkeysTable: QuadkeysTable,
    log: Log
  ) {
    this._elementTable = elementTable;
    this._quadkeysTable = quadkeysTable;
    this.log = log;
  }
  public getElementTable() {
    return this._elementTable;
  }
  public getQuadkeysTable() {
    return this._quadkeysTable;
  }
  public fork(newLog: Log, modifiedEntities: Entity[]) {
    if (newLog === this.log || newLog.length === 0) {
      return this;
    }
    // check if modified entities tally with latest log
    const latestEntry = newLog[0];

    if (
      latestEntry.size !== modifiedEntities.length ||
      modifiedEntities.some(r => !latestEntry.has(r.id))
    ) {
      throw new Error('log and modifiedEntities dont match');
    }

    modifiedEntities.forEach(e => {
      if (this._elementTable.has(e.id)) {
        throw new Error(`Modified entity ${e.id} already exists in table`);
      }
    });

    this.addVirgin(modifiedEntities, '');

    return new State(this._elementTable, this._quadkeysTable, newLog);
  }

  public getElement(id: string) {
    return this._elementTable.get(id);
  }

  public getIdByQuadkey(quadkey: string) {
    return this._quadkeysTable.get(quadkey);
  }

  public getLog() {
    return this.log;
  }

  public addVirgin(entities: Entity[], quadkey: string) {
    elementTableBulkAdd(this._elementTable, entities);

    quadkeysTableAdd(this._quadkeysTable, entities.map(r => r.id), quadkey);
  }

  public getVisible(quadkeys: string[]) {
    const insideQuadkeys = quadkeysTableFindVirginIds(
      this._quadkeysTable,
      quadkeys
    );

    const toRemoveIds = logVirginIdsCurrentlyModified(this.log);

    for (const id of toRemoveIds) {
      insideQuadkeys.delete(id);
    }

    const latestIndexes = getAllLatestIndexes(this.log);

    for (const id of latestIndexes) {
      insideQuadkeys.add(id);
    }

    return insideQuadkeys;
  }

  public shred(quadkey: string) {
    // To test we dont remove '' key
    const shreddedQuadkeyTable = quadkeysTableCreateFrom(
      this._quadkeysTable,
      quadkey
    );

    // gets all indexes
    const shreddedIndexes = quadkeysTableFlatten(shreddedQuadkeyTable);

    const shreddedElementTable = tableFilter(
      (_, id) => shreddedIndexes.has(id),
      this._elementTable
    );

    this._elementTable = shreddedElementTable;
    this._quadkeysTable = shreddedQuadkeyTable;
  }
}
