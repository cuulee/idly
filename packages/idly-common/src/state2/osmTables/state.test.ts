import * as R from 'ramda';
import {
  nodeFactory,
  relationFactory,
  wayFactory,
} from '../../osm/entityFactory/index';
import { Entity } from '../../osm/structures';
import { setCreate } from '../helper';
import { addEntryToLog, Log, logCreate } from '../log/index';
import {
  Element,
  elementTableBulkAdd,
  elementTableCreate,
} from './elementTable';
import { quadkeysTableAdd, quadkeysTableCreate } from './quadkeysTable';
import { State } from './state';

const mapFromObj = <T>(o: any = {}): Map<string, T> =>
  Object.keys(o).reduce((prev, k) => {
    prev.set(k, o[k]);
    return prev;
  }, new Map());

const dummyElement = (
  entity: Entity,
  parentWays = setCreate<string>(),
  parentRelations = setCreate<string>()
): Element => ({
  entity,
  parentRelations,
  parentWays,
});

const dummyElementT = (e: Entity[]) =>
  R.tap(t => elementTableBulkAdd(t, e), elementTableCreate());

const dummyQuadkeyT = (obj: any): any => mapFromObj(obj);

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});
const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

const w2 = wayFactory({
  id: 'w2',
  nodes: ['n1', 'n3'],
});

const r1 = relationFactory({
  id: 'r1',
  members: [{ id: 'n4', ref: 'n4' }, { id: 'w2', ref: 'w2' }],
});

describe('constructing state and basic tests', () => {
  it('should create', () => {
    expect(State.create()).toEqual({
      _elementTable: mapFromObj(),
      _quadkeysTable: mapFromObj(),
      log: [],
    });
  });
  it('should create with given values', () => {
    const elementTable = elementTableCreate();
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');
    const opt = {
      elementTable,
      log: [setCreate([])],
      quadkeysTable,
    };
    expect(State.create(opt)).toMatchSnapshot();
  });
});

describe('fork', () => {
  it('should work with blank arrays', () => {
    const elementTable = elementTableCreate();
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');

    const log = logCreate();

    expect(
      State.create({
        elementTable,
        log,
        quadkeysTable,
      }).fork([], [])
    ).toEqual(
      State.create({
        elementTable,
        log,
        quadkeysTable,
      })
    );
  });

  it('should return same ref if same log is passed', () => {
    const elementTable = elementTableCreate();
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');

    const log = logCreate();
    const state = State.create({
      elementTable,
      log,
      quadkeysTable,
    });

    expect(state.fork(log, [])).toBe(state);
  });

  it('should do a simple fork', () => {
    const elementTable = dummyElementT([n1, n2, w1]);
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');
    quadkeysTableAdd(quadkeysTable, ['n3'], '12210');

    const log1 = logCreate();

    const state1 = State.create({
      elementTable,
      log: log1,
      quadkeysTable,
    });

    const n1Hash0 = nodeFactory({
      id: 'n1#0',
    });

    const log2: Log = addEntryToLog(setCreate([n1Hash0.id]))(log1);

    const state2 = state1.fork(log2, [n1Hash0]);

    expect(state2).toEqual(
      State.create({
        elementTable: dummyElementT([n1, n2, w1, n1Hash0]),
        log: log2,
        quadkeysTable: dummyQuadkeyT({
          '': setCreate(['n1#0']),
          '12210': setCreate(['n3']),
          '123': setCreate(['n1', 'n2', 'w1']),
        }),
      })
    );
  });

  it('should throw error when log and modified entities dont match', () => {
    const elementTable = elementTableCreate();
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');

    const log1 = logCreate();
    const state = State.create({
      elementTable,
      log: log1,
      quadkeysTable,
    });
    const n1Hash0 = nodeFactory({
      id: 'n1#0',
    });
    const n2Hash0 = nodeFactory({
      id: 'n2#0',
    });

    const w1Hash0 = wayFactory({
      id: 'w1#0',
      nodes: [n1Hash0.id, n2Hash0.id],
    });

    const log2 = addEntryToLog(setCreate([n1Hash0.id, w1Hash0.id]))(log1);
    expect(() =>
      state.fork(log2, [n1Hash0, n2Hash0])
    ).toThrowErrorMatchingSnapshot();

    const log3 = addEntryToLog(setCreate([n1Hash0.id, w1Hash0.id]))(log1);
    expect(() => state.fork(log3, [n1Hash0])).toThrowErrorMatchingSnapshot();
  });
});

describe('getVisible', () => {
  it('should get ids', () => {
    const elementTable = dummyElementT([n1, n2, w1]);
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');

    const log1 = logCreate();

    const state1 = State.create({
      elementTable,
      log: log1,
      quadkeysTable,
    });
    state1.addVirgin([n3], '121');
    state1.addVirgin([n4], '130');

    const n1Hash0 = nodeFactory({
      id: 'n1#0',
    });
    const n2Hash0 = nodeFactory({
      id: 'n2#0',
    });
    const w1Hash0 = wayFactory({
      id: 'w1#0',
      nodes: [n1Hash0.id, n2Hash0.id],
    });

    const log2 = addEntryToLog(
      setCreate([n1Hash0, n2Hash0, w1Hash0].map(r => r.id))
    )(log1);

    const state2 = state1.fork(log2, [n1Hash0, n2Hash0, w1Hash0]);

    expect(state2.getVisible(['123'])).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0].map(e => e.id))
    );

    expect(state2.getVisible(['121'])).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0, n3].map(e => e.id))
    );

    expect(state2.getVisible(['130'])).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0, n4].map(e => e.id))
    );

    expect(state2.getVisible(['1'])).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0, n3, n4].map(e => e.id))
    );
  });
});

describe('addVirgin', () => {
  it('should add', () => {
    const elementTable = dummyElementT([n1, n2, w1]);
    const quadkeysTable = quadkeysTableCreate();
    elementTableBulkAdd(elementTable, [n1, n2, w1]);
    quadkeysTableAdd(quadkeysTable, ['n1', 'n2', 'w1'], '123');

    const log1 = logCreate();

    const state1 = State.create({
      elementTable,
      log: log1,
      quadkeysTable,
    });

    const n1Hash0 = nodeFactory({
      id: 'n1#0',
    });

    const n2Hash0 = nodeFactory({
      id: 'n2#0',
    });

    state1.addVirgin([n3, n4], '121');
    expect(state1.getElementTable()).toMatchSnapshot();
    expect(state1.getQuadkeysTable()).toMatchSnapshot();

    const log2 = addEntryToLog(setCreate([n1Hash0, n2Hash0].map(r => r.id)))(
      log1
    );

    const state2 = state1.fork(log2, [n1Hash0, n2Hash0]);
    expect(state2.getElementTable()).toMatchSnapshot();
    expect(state2.getQuadkeysTable()).toMatchSnapshot();

    state2.addVirgin([w2, r1], '33');
    expect(state2.getElementTable()).toMatchSnapshot();
    expect(state2.getQuadkeysTable()).toMatchSnapshot();

    const n3Hash0 = nodeFactory({
      id: 'n3#0',
    });

    const n1Hash1 = nodeFactory({
      id: 'n1#1',
    });

    const r1Hash0 = relationFactory({
      id: 'r1#0',
    });

    const log3 = addEntryToLog(
      setCreate([n3Hash0, n1Hash1, r1Hash0].map(r => r.id))
    )(log2);

    const state3 = state2.fork(log3, [n3Hash0, n1Hash1, r1Hash0]);

    expect([...state3.getElementTable().keys()]).toMatchSnapshot();
    expect(state3.getQuadkeysTable()).toMatchSnapshot();

    expect(state3.getVisible([''])).toMatchSnapshot();

    expect(state3.getVisible(['33'])).toMatchSnapshot();

    expect(state3.getVisible(['121'])).toMatchSnapshot();

    const r1Hash1 = relationFactory({
      id: 'r1#1',
    });

    const log4 = addEntryToLog(setCreate([r1Hash1].map(r => r.id)))(log3);
    const state4 = state3.fork(log4, [r1Hash1]);

    expect([...state4.getElementTable().keys()]).toMatchSnapshot();
    expect(state4.getVisible(['33']).has(r1Hash1.id)).toBe(true);
  });
});
