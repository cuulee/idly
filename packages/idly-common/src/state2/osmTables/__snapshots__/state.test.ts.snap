// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addVirgin should add 1`] = `
Map {
  "n1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "n2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "w1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n3" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n3",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n4" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n4",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
}
`;

exports[`addVirgin should add 2`] = `
Map {
  "123" => Set {
    "n1",
    "n2",
    "w1",
  },
  "121" => Set {
    "n3",
    "n4",
  },
}
`;

exports[`addVirgin should add 3`] = `
Map {
  "n1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "n2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "w1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n3" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n3",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n4" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n4",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n1#0" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1#0",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n2#0" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2#0",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
}
`;

exports[`addVirgin should add 4`] = `
Map {
  "123" => Set {
    "n1",
    "n2",
    "w1",
  },
  "121" => Set {
    "n3",
    "n4",
  },
  "" => Set {
    "n1#0",
    "n2#0",
  },
}
`;

exports[`addVirgin should add 5`] = `
Map {
  "n1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
      "w2",
    },
  },
  "n2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "w1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n3" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n3",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w2",
    },
  },
  "n4" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n4",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {
      "r1",
    },
    "parentWays": Set {},
  },
  "n1#0" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1#0",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n2#0" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2#0",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "w2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w2",
      "nodes": Array [
        "n1",
        "n3",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {
      "r1",
    },
    "parentWays": Set {},
  },
  "r1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "r1",
      "members": Array [
        Object {
          "id": "n4",
          "ref": "n4",
        },
        Object {
          "id": "w2",
          "ref": "w2",
        },
      ],
      "tags": Object {},
      "type": "relation",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
}
`;

exports[`addVirgin should add 6`] = `
Map {
  "123" => Set {
    "n1",
    "n2",
    "w1",
  },
  "121" => Set {
    "n3",
    "n4",
  },
  "" => Set {
    "n1#0",
    "n2#0",
  },
  "33" => Set {
    "w2",
    "r1",
  },
}
`;

exports[`addVirgin should add 7`] = `
Array [
  "n1",
  "n2",
  "w1",
  "n3",
  "n4",
  "n1#0",
  "n2#0",
  "w2",
  "r1",
  "n3#0",
  "n1#1",
  "r1#0",
]
`;

exports[`addVirgin should add 8`] = `
Map {
  "123" => Set {
    "n1",
    "n2",
    "w1",
  },
  "121" => Set {
    "n3",
    "n4",
  },
  "" => Set {
    "n1#0",
    "n2#0",
    "n3#0",
    "n1#1",
    "r1#0",
  },
  "33" => Set {
    "w2",
    "r1",
  },
}
`;

exports[`addVirgin should add 9`] = `
Set {
  "n3#0",
  "n1#1",
  "r1#0",
  "n2#0",
}
`;

exports[`addVirgin should add 10`] = `
Set {
  "w2",
  "n3#0",
  "n1#1",
  "r1#0",
  "n2#0",
}
`;

exports[`addVirgin should add 11`] = `
Set {
  "n4",
  "n3#0",
  "n1#1",
  "r1#0",
  "n2#0",
}
`;

exports[`addVirgin should add 12`] = `
Array [
  "n1",
  "n2",
  "w1",
  "n3",
  "n4",
  "n1#0",
  "n2#0",
  "w2",
  "r1",
  "n3#0",
  "n1#1",
  "r1#0",
  "r1#1",
]
`;

exports[`constructing state and basic tests should create with given values 1`] = `
State {
  "_elementTable": Map {
    "n1" => Object {
      "entity": Object {
        "attributes": Object {
          "changeset": undefined,
          "timestamp": undefined,
          "uid": undefined,
          "user": undefined,
          "version": undefined,
          "visible": undefined,
        },
        "id": "n1",
        "loc": Object {
          "lat": 0,
          "lon": 0,
        },
        "tags": Object {},
        "type": "node",
      },
      "parentRelations": Set {},
      "parentWays": Set {
        "w1",
      },
    },
    "n2" => Object {
      "entity": Object {
        "attributes": Object {
          "changeset": undefined,
          "timestamp": undefined,
          "uid": undefined,
          "user": undefined,
          "version": undefined,
          "visible": undefined,
        },
        "id": "n2",
        "loc": Object {
          "lat": 0,
          "lon": 0,
        },
        "tags": Object {},
        "type": "node",
      },
      "parentRelations": Set {},
      "parentWays": Set {
        "w1",
      },
    },
    "w1" => Object {
      "entity": Object {
        "attributes": Object {
          "changeset": undefined,
          "timestamp": undefined,
          "uid": undefined,
          "user": undefined,
          "version": undefined,
          "visible": undefined,
        },
        "id": "w1",
        "nodes": Array [
          "n1",
          "n2",
        ],
        "tags": Object {},
        "type": "way",
      },
      "parentRelations": Set {},
      "parentWays": Set {},
    },
  },
  "_quadkeysTable": Map {
    "123" => Set {
      "n1",
      "n2",
      "w1",
    },
  },
  "log": Array [
    Set {},
  ],
}
`;

exports[`fork should throw error when log and modified entities dont match 1`] = `"log and modifiedEntities dont match"`;

exports[`fork should throw error when log and modified entities dont match 2`] = `"log and modifiedEntities dont match"`;
